micro-rdk/src/common/data_manager.rs

```diff
@@ -0,0 +1,402 @@
+use std::time::Duration;
+
+use crate::common::data_collector::{DataCollectionError, DataCollector};
+use crate::common::data_store::DataStore;
+use crate::proto::app::data_sync::v1::SensorData;
+
+use super::data_collector::ResourceMethodKey;
+use super::data_store::{DataStoreError, WriteMode};
+use async_io::Timer;
+use bytes::BytesMut;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DataManagerError {
+    #[error("no data collectors in manager")]
+    NoCollectors,
+    #[error(transparent)]
+    CollectionError(#[from] DataCollectionError),
+    #[error(transparent)]
+    StoreError(#[from] DataStoreError),
+}
+
+pub struct DataManager<StoreType> {
+    collectors: Vec<DataCollector>,
+    store: StoreType,
+    sync_interval_ms: u64,
+    part_id: String,
+}
+
+impl<StoreType> DataManager<StoreType>
+where
+    StoreType: DataStore,
+{
+    pub fn new(
+        collectors: Vec<DataCollector>,
+        store: StoreType,
+        sync_interval_ms: u64,
+        part_id: String,
+    ) -> Self {
+        Self {
+            collectors,
+            store,
+            sync_interval_ms,
+            part_id,
+        }
+    }
+
+    pub fn sync_interval_ms(&self) -> u64 {
+        self.sync_interval_ms
+    }
+
+    pub fn part_id(&self) -> String {
+        self.part_id.clone()
+    }
+
+    pub(crate) fn collection_intervals(&self) -> Vec<u64> {
+        let mut intervals: Vec<u64> = self.collectors.iter().map(|x| x.time_interval()).collect();
+        intervals.sort();
+        intervals.dedup();
+        intervals
+    }
+
+    pub async fn run(&mut self) -> Result<(), DataManagerError> {
+        let min_interval = self
+            .collection_intervals()
+            .into_iter()
+            .min()
+            .ok_or(DataManagerError::NoCollectors)?;
+        let min_interval = std::cmp::min(min_interval, self.sync_interval_ms);
```

Comment:
Should there be some sort of floor for `min_interval`? Some sort of quantum that all intervals are snapped to to hopefully reduce the number of unique intervals if there are values that are close to the same?

Reply:
Yes, you're right. The logic in `collection_intervals` does not account for this. Will update